crates/gitbutler-repo/src/repository_ext.rs

```diff
@@ -634,3 +629,429 @@ impl GixRepositoryExt for gix::Repository {
         Ok(self)
     }
 }
+
+#[cfg(test)]
+mod test {
+    mod create_wd_tree {
+        use std::path::Path;
+
+        use crate::repository_ext::RepositoryExt as _;
+        use gitbutler_testsupport::testing_repository::{assert_tree_matches, TestingRepository};
+
+        /// These tests exercise the truth table that we use to update the HEAD
+        /// tree to match the worktree.
+        ///
+        /// Truth table for upsert/remove:
+        /// | HEAD Tree -> Index | Index -> Worktree | Action |
+        /// | add                | delete            | no-action |
+        /// | modify             | delete            | remove |
+        /// |                    | delete            | remove |
+        /// | delete             |                   | remove |
+        /// | delete             | add               | upsert |
+        /// | add                |                   | upsert |
+        /// |                    | add               | upsert |
+        /// | add                | modify            | upsert |
+        /// | modify             | modify            | upsert |
+        #[cfg(test)]
+        mod head_upsert_truthtable {
+            use super::*;
+
+            // | add                | delete            | no-action |
+            #[test]
+            fn index_new_worktree_delete() {
+                let test_repository = TestingRepository::open();
```

Comment:
Here it would be better to work with `git_testsupport::readonly::*` as it allows to create persistent read-only repositories that are faster to run on, while also being easy to introspect. I.e. one can enter the directory and manually validate that the repository is indeed in the state that it should be in, maybe even with a `git status --porcelain` to validate in the test itself to serve as 'show and validation of pre-condition'.